### ISSUE TITLE
Feature request: adaptive idle polling and optional weight unload for the serve loop

### ISSUE THREAD

**Issue by jroth1111** [waybarrios/vllm-mlx#508]
## Summary

Two related quality-of-life requests for `vllm-mlx serve` so the
OpenAI/Anthropic-compatible HTTP server can stay running indefinitely on
Apple Silicon without consuming appreciable CPU or memory between
requests:

1. **Light: adaptive idle polling.** Make `EngineConfig.step_interval`
   adaptive (or event-driven) so the async loop does not poll at 1 kHz
   when the scheduler is empty.
2. **Heavy: lazy load + idle unload.** Optional `--lazy-load` and
   `--idle-unload-seconds N` flags on `serve` to release model weights
   after a configurable idle timeout and reload on next request.

Goal is *minimal* idle footprint, not literally zero.

### 1 ms polling loop

`vllm_mlx/engine_core.py`:

```python
@dataclass
class EngineConfig:
    ...
    step_interval: float = 0.001  # 1ms between steps

while self._running:
    if self.scheduler.has_requests():
        output = await loop.run_in_executor(_executor, self.scheduler.step)
        ...
    else:
        await asyncio.sleep(step_interval)
```

With no requests in flight, the loop wakes ~1000×/s. `step_interval` is
not exposed as a CLI flag.

### Eager load, no unload

`server.load_model(...)` builds the global `BatchedEngine` before
`uvicorn.run(...)`. Lifespan startup calls `await _engine.start()`,
which loads weights through `load_model_with_fallback(...)` on the
dedicated `mlx-step` executor. `BatchedEngine.stop()` already tears
the engine down cleanly — but only on process exit.

`vllm_mlx/worker.py` carries the inherited vLLM `sleep`/`wake_up` stubs
that no-op with a comment that sleep mode is "not applicable for MLX
(unified memory)". That is correct for the vLLM page-GPU-to-CPU
semantics, but does not address the user-facing goal of releasing
weight RAM during long idle.

### Light tier (small PR, no API surface)

Replace the fixed `asyncio.sleep(step_interval)` idle branch with either
an asyncio `Event` woken by `add_request(...)`, or a two-speed interval
(`step_interval` while requests are running, `idle_step_interval` such
as 100 ms or 1 s when empty). Either should drop idle CPU by ~50–500×
without changing first-token latency.

### Heavy tier

```bash
vllm-mlx serve MODEL \
  --lazy-load \
  --idle-unload-seconds 1800
```

- `--lazy-load`: bind the HTTP server without loading weights.
- `--idle-unload-seconds N`: after N seconds with no active generation,
  save prefix cache if enabled, stop the engine, drop model references,
  and return to standby. `0` disables unload after first load.
- `/health/ready` distinguishes `standby`, `loading`, `ready`,
  `unloading`, `error`.
- `/v1/models` keeps returning the configured served model name in
  standby.

## Implementation notes

`_start_llm()` already loads the model on a dedicated `mlx-step`
ThreadPoolExecutor and hands the same executor to `AsyncEngineCore` —
this matters because `mlx-lm` 0.31+ binds the per-thread
`generation_stream` to whichever thread first touches it, and loading
on the asyncio loop thread leaks a stray `Stream(gpu, 1)` that breaks
later evals. Any lazy reload should re-enter the same `start()` path
behind a lifecycle lock instead of introducing a parallel load path.

Suggested shape: lifecycle manager holds configured params; inference
routes `await ensure_loaded()` under an async lock; concurrent first
requests await one shared load task; active-request counter prevents
unload mid-generation; idle timeout saves prefix cache, then
`engine.stop()`, then `mx.clear_cache()`.

## Why this matters

Allows `vllm-mlx serve` to run as an always-on local inference service
for editors and agents on memory-constrained Macs, without users having
to choose between "endpoint always available" and "model weights not
sitting in RAM."

## Related

A more detailed downstream version of this proposal lives in
[raullenchai/Rapid-MLX#265](https://github.com/raullenchai/Rapid-MLX/issues/265).

**Comment by janhilgard** [comment]
Good write-up. Some notes on feasibility:

### Light tier (adaptive idle polling)

The simplest path is a two-speed interval: keep 1ms while `has_requests()` is true, switch to ~100ms when idle. An `asyncio.Event` woken by `add_request()` would be cleaner but requires plumbing the event into `MLLMScheduler.add_request()` / `EngineCore.add_request()` — still doable without API changes.

Current idle CPU cost of the 1ms poll is negligible on Apple Silicon (the `asyncio.sleep(0.001)` path does no MLX work), but it does prevent the process from dropping to truly zero CPU. A PR for two-speed polling would be a welcome small improvement.

### Heavy tier (lazy load + idle unload)

This is architecturally more involved:
- `engine.stop()` + `mx.clear_cache()` would release weights, but `mlx` memory pools don't always return to OS immediately (depends on `mx.set_memory_limit` and Metal driver behavior)
- Reload path must go through the same `start()` → `bind_generation_streams()` path (as noted in the proposal)
- The lifecycle state machine (`standby → loading → ready → unloading → standby`) would need to propagate to `/health` and block concurrent requests during transitions

For reference, on our 256GB system with two models totaling ~47GB, idle memory is not a concern — but this would be valuable for 16-32GB machines running a single model.

### Recommendation

Light tier is a straightforward PR (< 50 LOC change to `engine_core.py`). Heavy tier is a larger design effort — would benefit from a draft PR to discuss the lifecycle lock/state machine approach.
